### PR TITLE
💬 #1189 Add new message object api

### PR DESCRIPTION
**Pull request by stephanevieira75**
### message
``` JSON
{
  "id": "uuidv1",
  "channel_id": "uuid",
  "thread_id": "",

  "created_at": number,

  "type": "message"     // Classic message
           | "event"   // Hidden events like key changes, or other system events
           | "thread",  // Thread entry, content will be the head message content
  "subtype": null
           | "application" //Message from a connector
           | "deleted" //Message deleted by user
           | "system", // Message from system (channel activity)

  "application_id": "uuid", //Additional information if message was sent by application
  "user_id": "uuid", //Required

  "edited": boolean,
  "text": "string", // Original text typed by user and used in notifications
                    // It is parsed as markdown by frontend
                    // Most messages will only use this format
  "blocks": [BlockObject], //Alternative nested json to replace text (if defined it replaces the text field)
  "files": [FileObject], //List of files attached to message
  "context": {}, //Hidden custom data for message, old hidden_data

  "title": "string", //Overide message title, old hidden_data.custom_title
  "picture": "string", //Overide message picture, old hidden_data.custom_icon

  "stats": {
    "answers": number, //Number of threaded messages under this one
  },

  "pinned_info": { //If message if pinned (or it is undefined)
    "pinned_by": "uuid",
    "pinned_at": number
  },

  "reactions": [
     {"name": "raised_hands", "users": ["uuid", "uuid"], "count": number}
  ]

}
```

### message_files
``` JSON
{
	"company_id": "uuid",
	"message_id": "uuid",
	"id": "uuid-v4",
	"metadata":  {
		"source": "drive",
		"external_id": "external file id"
   	}
}
```

### message_blocks

``` JSON
{
	"type": "section",
	"text": {
		"type": "mrkdwn",
		"text": "hello world :) azezaeza *zaezaea* `ezaezaeza`"
	}
}
```

### CODE REVIEW

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php [hunk 1]
```diff
@@ -478,8 +478,106 @@ public function setIncrementAtTime($increment_at_time)
...
+    public function getStringMessageType() {
+        //0 from user, 1 from application, 2 from system
+        switch($this->getMessageType()) {
+            case 0: return "message";
+            case 1: return "thread";
+            case 2: return "event";
+            default:
+                return "message";
+        }
+    }
```
Comment (RomaricMourgues) [inline twake/backend/core/src/Twake/Discussion/Entity/Message.php:490]: This is not exactly right it should be:
*$this->getMessageType() === 0*: type="message", subtype=null
*$this->getMessageType() === 1*: type="message", subtype="application"
*$this->getMessageType() === 2*: type="message", subtype="system"

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php [hunk 2]
```diff
@@ -478,8 +478,106 @@ public function setIncrementAtTime($increment_at_time)
...
+            "blocks" => $messageEntity->setBlocks($messageEntity),
+            "files" => $messageEntity->setFiles($messageEntity),
+            "context" => $messageEntity->getHiddenData(),
+            "title" => $messageEntity->getHiddenData()['custom_title'] ?: null,
+            "picture" => $messageEntity->getHiddenData()['custom_icon'] ?: null,
+            "stats" => Array(
+                "answers" => $messageEntity->getResponsesCount()
+            ),
+            "pinned_info" => Array(
+                "pinned_by" => ($messageEntity->getSender() ? $messageEntity->getSender()->getId() : null), // TODO
```
Comment (RomaricMourgues) [inline twake/backend/core/src/Twake/Discussion/Entity/Message.php:569]: You can remove the TODO

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php [hunk 3]
```diff
@@ -478,8 +478,106 @@ public function setIncrementAtTime($increment_at_time)
...
+                "pinned_at" => 0,
+            ),
+            "reactions" => $messageEntity->getReactions(), // TODO Change Reactions format
```
Comment (RomaricMourgues) [inline twake/backend/core/src/Twake/Discussion/Entity/Message.php:572]: You still need to change the format then ?

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php [hunk 4]
```diff
@@ -498,6 +596,7 @@ public function getAsArray()
             "content" => $this->getContent(),
             "user_specific_content" => $this->getUserSpecificContent(),
             "increment_at_time" => $this->getIncrementAtTime(),
+            "api_object" => $api_object,
```
Comment (RomaricMourgues) [inline twake/backend/core/src/Twake/Discussion/Entity/Message.php:599]: You need to mix the content of "api_object" with the main array, with something like:
return array_merge(Array(..), $api_object)

**Review by RomaricMourgues — CHANGES_REQUESTED** [review]

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php (hunk 3 above)
Comment (stephanevieira75) [reply twake/backend/core/src/Twake/Discussion/Entity/Message.php:572]: Reactions are currently formatted in a different way:
Capture d’écran de 2021-04-26 16-40-22

Maybe I could change that if you want ? 😄

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php (hunk 2 above)
Comment (stephanevieira75) [reply twake/backend/core/src/Twake/Discussion/Entity/Message.php]: Removed

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php (hunk 1 above)
Comment (stephanevieira75) [reply twake/backend/core/src/Twake/Discussion/Entity/Message.php:498]: Updated :)

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php (hunk 4 above)
Comment (stephanevieira75) [reply twake/backend/core/src/Twake/Discussion/Entity/Message.php]: Done 👍🏻

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php (hunk 3 above)
Comment (RomaricMourgues) [reply twake/backend/core/src/Twake/Discussion/Entity/Message.php:579]: Yes, it is part of the PR goal to change also this format :)

Comment (RomaricMourgues) [reply twake/backend/core/src/Twake/Discussion/Entity/Message.php:579]: So instead of having a list, just add the name parameter to the list

Comment (stephanevieira75) [reply twake/backend/core/src/Twake/Discussion/Entity/Message.php]: Updated :)

File: twake/backend/core/src/Twake/Discussion/Entity/Message.php [hunk 5]
```diff
@@ -478,9 +478,135 @@ public function setIncrementAtTime($increment_at_time)
...
+        $reactions = Array();
+
+        if($messageEntity->getReactions() == null) return $reactions;
+
+        $orignal_reactions_array = $messageEntity->getReactions();
+        
+        foreach($orignal_reactions_array as $key => $value){
+            $new_reaction_object= Array(
+                "name" => $key
+            );
```
Comment (RomaricMourgues) [inline twake/backend/core/src/Twake/Discussion/Entity/Message.php:510]: `                "name" => $value["name"] ?: $key`

Comment (stephanevieira75) [reply twake/backend/core/src/Twake/Discussion/Entity/Message.php:510]: Done

**Review by RomaricMourgues — APPROVED** [review]

**Pull request merged**